glean-core/python/glean/glean.py

```diff
@@ -169,17 +179,26 @@ def configuration(cls):
         return cls._configuration
 
     @classmethod
-    def reset(cls):
+    def _reset(cls):
         """
         Resets the Glean singleton.
         """
         # TODO: 1594184 Send the metrics ping
+
+        # WARNING: Do not run any tasks on the Dispatcher from here since this
+        # is called atexit, which also waits for the Dispatcher queue to
+        # complete.
+
         Dispatcher.reset()
         if cls._initialized:
             _ffi.lib.glean_destroy_glean()
         cls._initialized = False
         if cls._destroy_data_dir and cls._data_dir.exists():
-            shutil.rmtree(str(cls._data_dir))
+            # This needs to be run in the same one-at-a-time process as the
```

Comment:
I might be misunderstanding this, but the upload worker is notably a _different_ process, it's just that these processes are run one-by-one, right?

Reply:
That's correct.  I'll try to clarify this.  It's not *the* upload worker process (because we fire off a new one for each upload task), but only one runs at a time.